### PR TITLE
Dockerfile: add `ENTRYPOINT` directive

### PR DESCRIPTION
**Pull request by drupol**
Currently, running `docker run -it ghcr.io/typst/typst:latest` runs a root shell instead of running Typst.
I believe that the Typst container should only run Typst and nothing else.

Running a root shell inside a Docker image, instead of directly running the intended application (in this case, Typst), can present several risks and challenges, particularly around security and best practices. Here’s why it’s not advisable:

1. Security Risks: Running as root within a container, especially if it's interactive, significantly increases the risk of privilege escalation attacks.
2. Principle of Least Privilege: This principle advises that any process or user should have only the minimum privileges necessary to perform its function.
3. General ease of use: Using `docker run -it ghcr.io/typst/typst` runs Typst instead of running a shell, making the use of Typst within Docker is easier. 

As a workaround to use it, here's how I do... pretty ugly isn't it?: `docker run --entrypoint /bin/typst --mount type=bind,source="$(pwd)"/src,target=/src ghcr.io/typst/typst:latest compile /src/hello-world.typst`
I think there are room for improvements, and this is what this PR is fixing.

This PR fix this by:

1. Setting the `ENTRYPOINT` to `/bin/typst` by default
2. Hardening the image, do not use `root` user by default, use a dedicated user `typst`

I succeeded to build a project of mine by doing: 

```
docker build . -r typst:latest
docker run -v .:/home/typst -it typst:latest compile --root=. src/document/main.typ document1.pdf
```

### CODE REVIEW

**Comment by drupol** [comment]
Just tested it successfully!

```
❯ git co main
Switched to branch 'main'
Your branch is ahead of 'origin/main' by 39 commits.
  (use "git push" to publish your local commits)
❯ git pl
remote: Enumerating objects: 3256, done.
remote: Counting objects: 100% (1459/1459), done.
remote: Compressing objects: 100% (272/272), done.
remote: Total 3256 (delta 1295), reused 1270 (delta 1186), pack-reused 1797
Receiving objects: 100% (3256/3256), 4.47 MiB | 860.00 KiB/s, done.
Resolving deltas: 100% (2301/2301), completed with 370 local objects.
From github.com:drupol/typst
 + 09f8fa80...02683f62 patch-1    -> origin/patch-1  (forced update)
Current branch main is up to date.
❯ micro Dockerfile
❯ docker build .
[+] Building 492.3s (13/13) FINISHED                                                                                                                                                                                                                                                                                                  docker:default
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                                                                                                 0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                                                                            0.0s
 => => transferring dockerfile: 317B                                                                                                                                                                                                                                                                                                            0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                                                                                                                                                                                                                                                                0.4s
 => [internal] load metadata for docker.io/library/rust:alpine                                                                                                                                                                                                                                                                                  0.4s
 => [internal] load build context                                                                                                                                                                                                                                                                                                              21.6s
 => => transferring context: 6.85GB                                                                                                                                                                                                                                                                                                            21.6s
 => CACHED [build 1/4] FROM docker.io/library/rust:alpine@sha256:4a7925c3974fab3fb68b3f4d93d1a4a36cb201f9f87b01014def219b125f1960                                                                                                                                                                                                               0.0s
 => => resolve docker.io/library/rust:alpine@sha256:4a7925c3974fab3fb68b3f4d93d1a4a36cb201f9f87b01014def219b125f1960                                                                                                                                                                                                                            0.0s
 => [stage-1 1/3] FROM docker.io/library/alpine:latest@sha256:c5b1261d6d3e43071626931fc004f70149baeba2c8ec672bd4f27761f8e1ad6b                                                                                                                                                                                                                  0.0s
 => => resolve docker.io/library/alpine:latest@sha256:c5b1261d6d3e43071626931fc004f70149baeba2c8ec672bd4f27761f8e1ad6b                                                                                                                                                                                                                          0.0s
 => CACHED [stage-1 2/3] WORKDIR /root/                                                                                                                                                                                                                                                                                                         0.0s
 => [build 2/4] COPY . /app                                                                                                                                                                                                                                                                                                                    16.1s
 => [build 3/4] WORKDIR /app                                                                                                                                                                                                                                                                                                                    0.2s
 => [build 4/4] RUN apk add --update musl-dev     && cargo build -p typst-cli --release                                                                                                                                                                                                                                                       452.4s
 => [stage-1 3/3] COPY --from=build  /app/target/release/typst /bin                                                                                                                                                                                                                                                                             0.0s
 => exporting to image                                                                                                                                                                                                                                                                                                                          1.3s 
 => => exporting layers                                                                                                                                                                                                                                                                                                                         1.0s 
 => => exporting manifest sha256:bfcef7c38a42325b1e9e3454bd7df4eb404207e89fa648ddf71d188951c111fe                                                                                                                                                                                                                                               0.0s 
 => => exporting config sha256:97d7718f20f7aa89e4b3518bb6f2802e3f2b79d433d6e12ff5ed46cf5a865669                                                                                                                                                                                                                                                 0.0s 
 => => exporting attestation manifest sha256:57c7b9d88c10ee2511b57bc61308b720069b580b0ee1bca7e7accb1a2fd41d7c                                                                                                                                                                                                                                   0.0s 
 => => exporting manifest list sha256:9ec47a6761afc7658c360964b6fbfed0f0a5b285ef9eb2596670a3a8cd877e8e                                                                                                                                                                                                                                          0.0s
 => => naming to moby-dangling@sha256:9ec47a6761afc7658c360964b6fbfed0f0a5b285ef9eb2596670a3a8cd877e8e                                                                                                                                                                                                                                          0.0s
 => => unpacking to moby-dangling@sha256:9ec47a6761afc7658c360964b6fbfed0f0a5b285ef9eb2596670a3a8cd877e8e                                                                                                                                                                                                                                       0.3s
❯ docker images
REPOSITORY            TAG       IMAGE ID       CREATED          SIZE
<none>                <none>    9ec47a6761af   27 seconds ago   60.2MB
❯ docker run -it 9ec47a6761af
The Typst compiler

Usage: typst [OPTIONS] <COMMAND>

Commands:
  compile  Compiles an input file into a supported output format [aliases: c]
  watch    Watches an input file and recompiles on changes [aliases: w]
  query    Processes an input file to extract provided metadata
  fonts    Lists all discovered fonts in system and custom font paths
  update   Self update the Typst CLI (disabled)
  help     Print this message or the help of the given subcommand(s)

Options:
      --color[=<WHEN>]  Set when to use color. auto = use color if a capable terminal is detected [default: auto] [possible values: auto, always, never]
      --cert <CERT>     Path to a custom CA certificate to use when making network requests [env: TYPST_CERT=]
  -h, --help            Print help
  -V, --version         Print version
~/C/t/typst > main +1  [!>] > rust ✘ 
```

**Comment by alerque** [comment]
> 1. Security Risks: Running as root within a container, especially if it's interactive, significantly increases the risk of privilege escalation attacks.

I agree this PR is an improvement and a default entry point should be set. But **this is not a reason for that**. Setting the default entry point does *nothing* for the security of the container, that is something that can be overridden at runtime. Whether or not the container runs with root access is entirely up to the calling environment.

Go ahead with this change, but don't pretend it improves security. Everybody with permission to run the docker container will still have access to the root shell in the container even with this change.

**Comment by drupol** [comment]
That's definitely true, I've created a specific user `typst` to improve security.

File: Dockerfile [hunk 1]
```diff
@@ -7,5 +7,9 @@ RUN apk add --update musl-dev openssl-dev openssl-libs-static \
     cargo build -p typst-cli --release
 
 FROM alpine:latest
-WORKDIR /root/
+MAINTAINER Typst <hello@typst.app>
 COPY --from=build  /app/target/release/typst /bin
+WORKDIR /home/typst
+RUN adduser -D typst
+USER typst
```
Comment (alerque) [inline Dockerfile:14]: The entrypoint change is useful, but setting the user is not. It is only a hindrance to usefulness not a security improvement. Access to the container with root permissions is still possible anyway, and now you've hindered a workflow where the calling envirenment can easily set the user ID to match the UID/GID of the user on the host system so that any volumes mounted inside the container (for example a project directory) can be read/written as the same user that owns the files instead of some arbitrary UID from the container that won't even be readable by the user that wrote them.

Comment (drupol) [reply Dockerfile:14]: Let me double check that later in the day, I left computer now :S

Comment (drupol) [reply Dockerfile]: I removed the user creation since accessing the root container was still possible indeed. let me know if it's better.

Comment (alerque) [reply Dockerfile]: It's hard to review from mobile but from what I can see that looks better. Not sure on Alpine default paths but if they prefer system package stuff in /usr and 3rd party source built stuff in /usr/local like many distros do and if the latter is in the default path you should consider putting the bin there.

Comment (drupol) [reply Dockerfile]: Mmh I don't think moving it to `/usr/local/bin/` or `/usr/bin/` is going to make a difference. Is it really needed?

Comment (alerque) [reply Dockerfile]: Needed? No of course not it works either way, but in general the closer you follow distribution expectations the less surprises there will be down the road for end users.

Comment (drupol) [reply Dockerfile]: Sure, but in this case, it follows FHS and I don't think moving it to one or the other place will make a difference.

**Comment by alerque** [comment]
Besides the possible nitpick of Alpine's filesystem layout guidelines (which I haven't looked up, I just know this doesn't jive with the FHS spec) this looks fine now.

**Comment by drupol** [comment]
By the way, don't you think we should fix tag to a fixed commit ID instead of using `latest`?

**Comment by alerque** [comment]
> By the way, don't you think we should fix tag to a fixed commit ID instead of using `latest`?

Yes for sure. Or at least accept at ARG that defaults to latest but allows building from a specific tagged base container for reproducibility.

**Comment by drupol** [comment]
OK going to work on that.

**Comment by laurmaedje** [comment]
There are a few conflicts here now due to https://github.com/typst/typst/pull/3627 being merged.

**Comment by drupol** [comment]
I've fixed the conflicts, but let me carefully test this tomorrow.

It would be nice to have inputs from @alerque and @freundTech too.

File: Dockerfile [hunk 2]
```diff
@@ -22,6 +25,16 @@ RUN --mount=type=cache,target=/root/.cargo/git/db \
     cp target/$(xx-cargo --print-target-triple)/release/typst target/release/typst && \
     xx-verify target/release/typst
 
-FROM alpine:latest
-WORKDIR /root/
+FROM alpine:3.19
+LABEL org.opencontainers.image.authors="Typst <hello@typst.app>"
+LABEL org.opencontainers.image.created=${BUILD_DATE}
+LABEL org.opencontainers.image.description="Typst is an opensource markup-based typesetting system"
```
Comment (reknih) [inline Dockerfile:31]:
```suggestion
LABEL org.opencontainers.image.description="A markup-based typesetting system"
```

File: Dockerfile [hunk 3]
```diff
@@ -22,6 +25,16 @@ RUN --mount=type=cache,target=/root/.cargo/git/db \
...
+LABEL org.opencontainers.image.documentation="https://typst.app/docs"
+LABEL org.opencontainers.image.licenses="Apache-2.0"
+LABEL org.opencontainers.image.revision=${REVISION}
```
Comment (reknih) [inline Dockerfile:34]: Why did you add this but not `org.opencontainers.image.version` which I feel is more relevant?

File: Dockerfile [hunk 4]
```diff
@@ -22,6 +25,16 @@ RUN --mount=type=cache,target=/root/.cargo/git/db \
...
+LABEL org.opencontainers.image.source="https://github.com/typst/typst"
+LABEL org.opencontainers.image.title="Typst Docker image"
+LABEL org.opencontainers.image.url="https://typst.app"
+LABEL org.opencontainers.image.vendor="Typst"
```
Comment (reknih) [inline Dockerfile:38]:
```suggestion
LABEL org.opencontainers.image.vendor="Typst"

```

File: Dockerfile [hunk 5]
```diff
@@ -22,6 +25,16 @@ RUN --mount=type=cache,target=/root/.cargo/git/db \
     cp target/$(xx-cargo --print-target-triple)/release/typst target/release/typst && \
     xx-verify target/release/typst
 
-FROM alpine:latest
-WORKDIR /root/
+FROM alpine:3.19
+LABEL org.opencontainers.image.authors="Typst <hello@typst.app>"
```
Comment (reknih) [inline Dockerfile:29]: This should be the same name as in the `Cargo.toml`, you can, however, keep the email address

File: .github/workflows/docker-image.yml [hunk 6]
```diff
@@ -32,6 +32,9 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v4
 
+      - name: Get current date
+        run: echo "BUILD_DATE=\"$(date -u +'%Y-%m-%dT%H:%M:%SZ')\" >> ${GITHUB_ENV}"
```
Comment (reknih) [inline .github/workflows/docker-image.yml:36]: We usually start environment variables with the prefix `TYPST_` to prevent collisions with other applications.

File: Dockerfile [hunk 7]
```diff
@@ -1,5 +1,9 @@
+ARG BUILD_DATE
+ARG REVISION
+ARG TARGETPLATFORM
+
 FROM --platform=$BUILDPLATFORM tonistiigi/xx AS xx
-FROM --platform=$BUILDPLATFORM rust:alpine AS build
+FROM --platform=$BUILDPLATFORM rust:alpine3.19 AS build
```
Comment (reknih) [inline Dockerfile:6]: Why did you choose this particular tag granularity instead of `alpine` (like before) or a digest (for reproducability)? Is the aim to prevent breakage by non-patch updates? If so, why is the Rust version not specified here?

File: Dockerfile (hunk 3 above)
Comment (drupol) [reply Dockerfile:34]: True, but I don't think I have access to that variable in the CI.

File: Dockerfile (hunk 7 above)
Comment (drupol) [reply Dockerfile:6]: Simply because I'm not aiming for reproducibility here. If I really want to aim for reproducibility, I would build the image with Nix.

Comment (reknih) [reply Dockerfile:6]: I'd still be more comfortable if we included the Rust version in the tag. Our closed-source Dockerfiles start with a `RUST_VERSION` arg at the top which is then interpolated in the `FROM` statement:

```Dockerfile
ARG RUST_VERSION=1.74.0
# ...
FROM rust:${RUST_VERSION}-slim-bookworm AS build-bundler
```

If you decide to add this, you should also add a comment, both in the Dockerfile and Cargo.toml that advises to keep this in sync with `workspace.package.rust-version`.

File: Dockerfile (hunk 3 above)
Comment (reknih) [reply Dockerfile:34]: Don't you do via [`${{ github.event.release.tag_name }}`](https://stackoverflow.com/a/59521156/1567835)? This is run in a release workflow after all. You can add it as a build arg, just like `REVISION`

File: Dockerfile (hunk 7 above)
Comment (laurmaedje) [reply Dockerfile:6]: Note that `workspace.package.rust-version` is the _minimum_ supported Rust version, not the one we mainly use.

**Comment by laurmaedje** [comment]
what's the state of this?

**Comment by drupol** [comment]
At the beginning of this PR, the idea was just to add an `ENTRYPOINT`.

Then along the way I made other modifications leading to changes in the CI.

Are you OK if I just add the `ENTRYPOINT` within this PR and possibly make further improvements in another PR? I don't feel so comfortable with `Dockerfile` tweaking yet.

**Comment by alerque** [comment]
Most or all of these changes are probably okay, but it is a bit hard to reason about them all at once. The base image version pinning for example has pros and cons. The meta data tagging is probably just good, but then that brings in CI changes and it it becomes unclear what is related to what. I thing breaking this down in steps for discussion would make it much easier.

The ENTRYPOINT change by itself is pretty much a no-brainier improvement.

**Comment by drupol** [comment]
OK I'll make a separate PR for the metadata stuff. I'm not on the computer right now, I'll do that ASAP.

**Comment by drupol** [comment]
@alerque I've updated it, I don't think we need 2 PRs for those changes, let me know if it's clearer.

**Comment by reknih** [comment]
Thank you!
